test/basic_store.cpp

```diff
@@ -32,5 +38,230 @@ static_assert( std::is_move_constructible   <pool>{}, "");
+                return;
+            ec = {};
+        }
+        // Insert and Fetch
+        if(keySize > 1)
+        {
+            for(std::size_t n = 0; n < N; ++n)
+            {
+                auto item = ts[n];
+                ts.db.fetch(item.key,
+                    [&](void const* data, std::size_t size)
+                    {
+                        if(! BEAST_EXPECT(size == item.size))
+                            return;
+                        BEAST_EXPECT(
+                            std::memcmp(data, item.data, size) == 0);
+                    }, ec);
+                item = ts[N + n];
+                ts.db.insert(item.key, item.data, item.size, ec);
+                if(! BEAST_EXPECTS(! ec, ec.message()))
+                    return;
+                ts.db.fetch(item.key,
+                    [&](void const* data, std::size_t size)
+                    {
+                        if(! BEAST_EXPECT(size == item.size))
+                            return;
+                        BEAST_EXPECT(
+                            std::memcmp(data, item.data, size) == 0);
+                    }, ec);
+            }
+        }
+        if(sleep)
+        {
+            // Trigger a shrink_to_fit
+            std::this_thread::sleep_for(
+                std::chrono::milliseconds{2000});
+        }
+        ts.close(ec);
+        if(! BEAST_EXPECTS(! ec, ec.message()))
+            return;
+    }
+
+    // Perform insert/fetch test across a range of parameters
+    void
+    test_insert_fetch()
+    {
+        for(auto const keySize : {
+            1, 2, 3, 31, 32, 33, 63, 64, 65, 95, 96, 97 })
+        {
+            std::size_t N;
+            std::size_t constexpr blockSize = 4096;
+            float loadFactor = 0.95f;
+            switch(keySize)
+            {
+            case 1: N = 10; break;
+            case 2: N = 100; break;
+            case 3: N = 500; break;
+            default:
+                N = 5000;
+                break;
+            };
+            do_insert_fetch(N, keySize, blockSize, loadFactor,
+                keySize == 97);
+        }
+    }
+
+    template<class F>
+    void
+    check_throw(F const& f)
+    {
+        try
+        {
+            f();
+            BEAST_EXPECT(false);
+        }
+        catch(std::exception const&)
+        {
+            pass();
+        }
+        catch(...)
+        {
+            BEAST_EXPECT(false);
+        }
+    }
+
+    // 32-bit std::size_t
+    template<class = void>
+    void
+    test_throws(test_store&, std::false_type)
+    {
+    }
+
+    // 64-bit std::size_t
+    template<class = void>
+    void
+    test_throws(test_store& ts, std::true_type)
+    {
+        static_assert(sizeof(std::size_t) >= 8, "");
+        std::size_t size =
+            std::numeric_limits<std::size_t>::max() + 1;
+        error_code ec;
+        check_throw([&]{ ts.db.insert(nullptr, nullptr, size, ec); });
+    }
```

Comment:
Consider testing the size zero case
check_throw([&]{ ts.db.insert(nullptr, nullptr, 0, {}); });